File: NeetCode 150/Trie/211. Design Add and Search Words Data Structure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.endofWord = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()

            node = node.children[ch]
        node.endofWord = True

        

    def search(self, word: str) -> bool:
        def dfs(node,index):
            if index == len(word):
                return node.endofWord
            
            ch = word[index]
            if ch == '.':
                for child in node.children.values():
                    if dfs(child,index+1):#since I am searching for one true case. Then break the loop and return
                        return True
                return False #if no child is true, return False
                        
            else:
                if ch not in node.children:
                    return False
                return dfs(node.children[ch],index+1)
        return dfs(self.root,0)

        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: NeedCode 150 (JAVA)/Trie/211. Design Add and Search Words Data Structure.java
class WordDictionary {


    private class TrieNode{
        TrieNode[] children = new TrieNode[26];
        boolean endofWord;
    }
     
    private TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode node = root;

        for(char c : word.toCharArray()){
            int idx = c - 'a';
            if(node.children[idx] == null)
                node.children[idx] = new TrieNode();
            node = node.children[idx];
        }
        node.endofWord = true;
        
    }
    
    public boolean search(String word) {
        return dfs(word,0,root);
    }

    private boolean dfs(String word, int index, TrieNode node){
        if (node == null) return false;
        if (word.length() == index) return node.endofWord;

        char c = word.charAt(index);
        if(c =='.'){
            for(TrieNode child: node.children){
                if (dfs(word, index+1, child))
                    return true;
            }
            return false;
        }
        else{
            int idx = c - 'a';
            return dfs(word, index+1, node.children[idx]);
        }
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
